packages/debug/src/browser/abstract-launch-preference-provider.ts

```diff
@@ -0,0 +1,157 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, postConstruct } from 'inversify';
+import { PreferenceScope, PreferenceProvider } from '@theia/core/lib/browser';
+import { Emitter, Event } from '@theia/core';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { DisposableCollection } from '@theia/core';
+import { Disposable } from '@theia/core';
+
+export interface GlobalLaunchConfig {
+    version: string;
+    compounds?: LaunchCompound[];
+    configurations: LaunchConfig[];
+}
+
+export namespace GlobalLaunchConfig {
+    /* tslint:disable-next-line:no-any */
+    export function is(data: any): data is GlobalLaunchConfig {
+        return !data || (!!data.version && (!data.compounds || Array.isArray(data.compounds)) && Array.isArray(data.configurations));
+    }
+}
+
+export interface LaunchConfig {
+    type: string;
+    request: string;
+    name: string;
+
+    /* tslint:disable-next-line:no-any */
+    [field: string]: any;
+}
+
+export interface LaunchCompound {
+    name: string;
+    configurations: (string | { name: string, folder: string })[];
+}
+
+export const LaunchPreferenceProvider = Symbol('LaunchConfigurationProvider');
+export interface LaunchPreferenceProvider {
+
+    readonly onDidLaunchChanged: Event<void>;
+
+    ready: Promise<void>;
+
+    getConfigurationNames(withCompounds: boolean, resourceUri?: string): string[];
+
+}
+
+export const FolderLaunchProviderOptions = Symbol('FolderLaunchProviderOptions');
+export interface FolderLaunchProviderOptions {
+    folderUri: string;
+}
+
+export const LaunchProviderProvider = Symbol('LaunchProviderProvider');
+export type LaunchProviderProvider = (scope: PreferenceScope) => LaunchPreferenceProvider;
+
+@injectable()
+export abstract class AbstractLaunchPreferenceProvider implements LaunchPreferenceProvider, Disposable {
+
+    protected readonly onDidLaunchChangedEmitter = new Emitter<void>();
+    readonly onDidLaunchChanged: Event<void> = this.onDidLaunchChangedEmitter.event;
+
+    protected preferences: GlobalLaunchConfig | undefined;
+
+    protected _ready: Deferred<void> = new Deferred<void>();
+
+    protected readonly toDispose = new DisposableCollection();
+
+    protected readonly preferenceProvider: PreferenceProvider;
+
+    @postConstruct()
+    protected init(): void {
+        this.preferenceProvider.ready
+            .then(() => this._ready.resolve())
+            .catch(() => this._ready.resolve());
+
+        this.updatePreferences();
+        if (this.preferences !== undefined) {
+            this.emitLaunchChangedEvent();
+        }
+
+        this.toDispose.push(this.onDidLaunchChangedEmitter);
+        this.toDispose.push(
+            this.preferenceProvider.onDidNotValidPreferencesRead(prefs => {
+                if (!prefs || !GlobalLaunchConfig.is(prefs.launch)) {
```

Comment:
Why should it go here and cannot happen in launch preference provider? then we don't  need a new event `onDidNotValidPreferencesRead` and can do `GlobalLaunchConfig.is` check in one place.

Reply:
> Why should it go here and cannot happen in launch preference provider? then we don't need a new event `onDidNotValidPreferencesRead` and can do `GlobalLaunchConfig.is` check in one place.

Event `onDidNotValidPreferencesRead` triggers creating new preference schema for launch property, so I don't see how we can't get rid of it.

> looks like if you don't use the validation function aware of setting schema for launch folder preferences maybe you won't get some false possible validation results and getting rid of an event would be easier.

I'm not sure I understood what you meant... Can you elaborate on this?

Reply:
Why do we need to create preferences from invalid data? If it is valid then it should be part of `onDidPreferencesChanged`. Could you elaborate on an example when preferences are not valid but we still want to use them?

Reply:
It's because how launch preference schema is defined.

https://github.com/theia-ide/theia/blob/397f360422dae3dc590e58665fd526d04d7810c7/packages/debug/src/browser/debug-schema-updater.ts#L186-L189

In order to build this schema I need to get all configuration names to fill the `enum`. On the other hand, to get launch configurations they need to be validated against launch preference schema which is not built yet.

Imagine, you have launch configuration 

```{
	"launch": {
		"confugurations": [
			{ "name": "Launch Backend", ...},
			{ "name": "Launch Frontend", ...}
		],
		"compounds": [
			{
				"name": "Compound",
				"configurations": [
					"Launch Backend", // valid
					"Launch Frontend", // valid
					"Run Tests" // not valid
				]
			}
		]
```

And you need to restrict possible values of `configurations` in `compounds` section to only available configuration names.

Reply:
Do you mean if someone add `Run Tests` to `compounds.configurations` then the whole `launch` preference marked as invalid and not available to a user?
Do you also mean that we reparse launch json twice, once to create a schema and second time to validate data? 

How about having custom validation for launch preferences to exclude only bogus compound configuration, break a cycle between schema and data, avoid parsing data twice and introducing a new event? I don't think you need to check each property in this case of each configuration just top-level structure, like here https://github.com/akurinnoy/theia/blob/397f360422dae3dc590e58665fd526d04d7810c7/packages/debug/src/browser/debug-configuration-model.ts#L74

Reply:
> Do you mean if someone add `Run Tests` to `compounds.configurations` then the whole `launch` preference marked as invalid and not available to a user?  
Do you also mean that we reparse launch json twice, once to create a schema and second time to validate data?

yes, you're correct. But we could cache invalid properties to avoid re-parsing launch.json for second time.

> How about having custom validation for launch preferences to exclude only bogus compound configuration, break a cycle between schema and data, avoid parsing data twice and introducing a new event? I don't think you need to check each property in this case of each configuration just top-level structure, like here

In this case, if a user gets or inspects `launch` using plugin API he will lose default value of it. Also, preference editor will treat `launch` property as not valid.

Reply:
I agree with Anton tha the schema should not depend on the content of the validated file. If json schema does not offer this capability, we should not validate or work around it.

Reply:
If I understand correctly, the thing that is not translatable to a json schema is the choice of launch configs in the "compound" section (needs to be one of those mentioned in "launch"). If we do not represent this relationship in the schema, what default values would that remove?

Reply:
If we don't represent that relationship, the schema becomes broken and couldn't be used to provide default values nor to validate configs.

Reply:
@akosyakov what do you think about defining  `compounds` as empty schema instead of filling it with launch configs, i.e.

```
schema.properties.compounds = {}
```

This allows to contribute the schema, thus validate only `configurations` section of launch configs, and to avoid everithing you've mentioned above.

Reply:
> If we don't represent that relationship, the schema becomes broken and couldn't be used to provide default values nor to validate configs.

I don't think so: the only thing that would disappear would be the validation of the choice in the "compuound" thingy. The set of all possible launch configurations (and their schemas) does not depend on the content of settings.json. You can still build a schema for each type of launch configuration.

Reply:
> In this case, if a user gets or inspects launch using plugin API he will lose default value of it. Also, preference editor will treat launch property as not valid.

Could you elaborate when default values are provided? I've tested VS Code api for `launch` and don't see any default values: https://github.com/akosyakov/vscode-launch/blob/master/src/test/extension.test.ts

Also I've learned that VS Code:
- returns content of launch section as it is, it only ensures that `configurations` and `compound` section exist and they are of Array type, but elements can be bogus
- `launch` is never empty even if there is no launch.json and settings.json and always has `compound` and `configurations` of an array type
- if launch.json provided and settings.json has `launch` section then the latter is ignored

There is a way to run tests from VS Code extension against Theia. We should do it for this PR to check that implementation is correct. Sample config: https://github.com/theia-ide/theia/issues/4224#issuecomment-459752557

Reply:
@akosyakov Sorry for late response (I was on another issue and then on a short vacation)

> Could you elaborate when default values for launch are provided?

Default value for launch property you've tested [here](https://github.com/akosyakov/vscode-launch/blob/master/src/test/extension.test.ts#L43). It's provided when no launch.json nor global launch configuration exist in a project.

> * ensures that `configurations` and `compound` section exist and they are of Array type, but elements can be bogus

Thanks to your extension I've figured out that VS Code ensures only exsistanse of `configurations` and `compounds` but not their type.
If `settings.json` has `launch` section and `launch.json` doesn't contain `configurations` and/or `compounds` then corresponding sub-section will be taken from `settings.json`.

Taking into account the above, `launch` configuration property shouldn't be verified. We should only enshure that `launch` has `configurations` and `compounds`, or take them from user settings or default values.

Reply:
I see, added inspect tests to check it: https://github.com/akosyakov/vscode-launch/blob/3da1365c015116484c0829da8b9468a44a2c05ed/src/test/extension.test.ts#L266-L317

I am not sure thought it's schema based, you can see from tests that the default value should be provided for `globalValue` (user settings) during inspect.

Reply:
I wonder whether launch launch configs in user settings makes sense at all? Should not we just stub it with default values? [Docs](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration) does not say anything about user settings:
> Note: Workspace and Workspace Folder configurations contains launch and tasks settings.

Reply:
Have you seen [default settings](https://code.visualstudio.com/docs/getstarted/settings#_default-settings) and explanation near `launch` property?

Reply:
I think launch config in user settings shouldn't be stabbed, and Theia should use a global launch config if a project doesn't have its own.

Reply:
Does it mean that it's stubbed with such value and data from user settings json are ignored? Do I read it correctly?

Reply:
I'm wrong. My user settings already had launch section that's why i got such results. If i try to changes it changes are reflected into expectations, so we should read launch section user settings as well.

I will update tests with expectations that user settings does not contain launch section.